uqcsbot/scripts/bgg.py

```diff
@@ -0,0 +1,180 @@
+    """
+    returns the bgg id, searching by name
+    """
+    query = get(f"https://www.boardgamegeek.com/xmlapi2/"
+                + f"search?type=boardgame,boardgameexpansion&query={search_name:s}")
+    if query.status_code != 200:
+        return None
+    results = fromstring(query.text)
+    if results.get("total", "0") == "0":
+        return None
+
+    # filters for the closest name match
+    match = {}
+    for element in results:
+        if element.get("id") is None:
+            continue
+        for subelement in element:
+            if subelement.tag == "name":
+                match[element.get("id")] = SequenceMatcher(None, search_name,
+                                                           subelement.get("value")).ratio()
+    return max(match, key=match.get)
+
+
+def get_board_game_parameters(identity: str) -> Optional[dict]:
+    """
+    returns the various parameters of a board game from bgg
+    """
+    query = get(f"https://www.boardgamegeek.com/xmlapi2/thing?stats=1&id={identity:s}")
+    if query.status_code != 200:
+        return None
+    result = fromstring(query.text)[0]
+    parameters: Dict[str, Any] = {}
+    parameters["categories"] = set()
+    parameters["mechanics"] = set()
+    parameters["subranks"] = {}
+    parameters["identity"] = identity
+
+    for element in result:
+        tag = element.tag
+        tag_name = element.attrib.get("name")
+        tag_value = element.attrib.get("value")
+        tag_type = element.attrib.get("type")
+        tag_text = element.text
+
+        # sets the range of players
+        if tag == "poll" and tag_name == "suggested_numplayers":
+            players = set()
+            for subelement in element:
+                numplayers = subelement.attrib.get("numplayers")
+                votes = 0
+
+                for subsubelement in subelement:
+                    numvotes = int(subsubelement.attrib.get("numvotes"))
+                    direction = -1 if subsubelement.attrib.get("value") == "Not Recommended" else 1
+                    votes += numvotes * direction
+
+                if votes > 0:
+                    players.add(numplayers)
+
+            if players:
+                parameters["min_players"] = min(players)
+                parameters["max_players"] = max(players)
+
+        # sets the name of the board game
+        elif tag == "name" and tag_type == "primary":
+            parameters["name"] = tag_value
+
+        # adds a category
+        elif tag == "link" and tag_type == "boardgamecategory":
+            parameters["categories"].add(tag_value)
+
+        # adds a mechanic
+        elif tag == "link" and tag_type == "boardgamemechanic":
+            parameters["mechanics"].add(tag_value)
+
+        # sets the user ratings
+        elif tag == "statistics":
+            for subelement in element[0]:
+                subtag = subelement.tag
+                subvalue = subelement.attrib.get("value")
```

Comment:
subelement and others should have more meaningful names (e.g. stat_element, stat_tag, stat_value) since these are statistics, not any element.

Reply:
Updated.